tests/lib/pkgdb.sh

```diff
@@ -893,3 +893,75 @@ distro_upgrade() {
             ;;
     esac
 }
+
+distro_list_packages() {
+    where="$1"
+
+    case "$SPREAD_SYSTEM" in
+        fedora-*|centos-*|amazon-linux-*|opensuse-*)
+            rpm -qa | sort > "$where"
+            ;;
+        ubuntu-*|debian-*)
+            dpkg --get-selections > "$where"
+            ;;
+        arch-*)
+            pacman -Q | cut -f1 -d' ' | sort > "$where"
+            ;;
+        *)
+            ;;
+    esac
+}
+
+distro_restore_packages() {
+    before="$1"
+    after="$before.after"
+    installed="$before.installed"
+    removed="$before.removed"
+
+    # find out what's installed now on systems where the old list is not enough
+    distro_list_packages "$after"
+    # make sure that the package list is not empty
+    test -s "$after"
+
+    if diff -up "$before" "$after"; then
+        # system pacakges did not change, nothing to do
+        return 0
+    fi
+
+    grep -v -f "$before" "$after" > "$installed" || true
```

Comment:
This warrants a comment. I would honestly use python for this, as it's a 3-4 liner and more readable and safe. I think you need at leas `-F` and `-x` to make this correct though.

Reply:
Actually it's a one liner:

```
any-python -c '"\n".join(print(set(open("after").read().splitlines()) - set(open("before").read().splitlines())))' > "$installed"
any-python -c '"\n".join(print(set(open("before").read().splitlines()) - set(open("after").read().splitlines())))' > "$removed"
```

Use `any-python`

Reply:
that's a pretty long one-liner.... I agree this would be cleaner in python but I don't think we get more readable code by sticking a 121 character long line of python that calls 8 functions into this shell script. It might be more correct but I don't think it's more readable.

Reply:
I think the meaning of the grep has yet to be ascertained. I was just guessing what maciek meant.